domain/model/modules.py

```diff
@@ -461,4 +358,236 @@ def __init__(
+        """
+
+        super(VisionTransformer, self).__init__()
+
+        self.input_shape = input_shape
+        self.in_channels = in_channels
+        self.num_classes = num_classes
+        self.dropout_rate = dropout_rate
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.num_attention_heads = num_attention_heads
+        self.num_layers = num_layers
+        self.attention_head_dim = attention_head_dim
+        self.mlp_head_dim = mlp_head_dim
+        self.freq = freq
+        
+        self.patch_embedding = PatchEmbedding(input_shape, in_channels, patch_size, embed_dim, freq)
+
+        self.num_patches = self.patch_embedding.num_patches
+
+        self.dropout_layer = nn.Dropout(dropout_rate)
+
+        self.cls_token = nn.Parameter(torch.randn(1, 1, self.embed_dim))
+
+        self.positional_embedding=nn.Parameter(torch.randn(1, self.num_patches, self.embed_dim))
+
+        self.transformer_encoder = TransformerEncoder(
+                                num_attention_heads=num_attention_heads, 
+                                num_layers=num_layers,
+                                embed_dim=512, 
+                                attention_head_dim=attention_head_dim,
+                                mlp_head_dim=mlp_head_dim,
+                                dropout_rate=dropout_rate)
+
+        self.prediction_mlp=nn.Sequential(nn.LayerNorm(self.embed_dim), nn.Linear(self.embed_dim, num_classes))
+
+
+
+    def forward(self, x):
+        batch_size = x.size(0)
+        embedding, num_patches = self.patch_embedding(x)
+
+        x = torch.cat((self.cls_token.repeat(batch_size, 1, 1), embedding), dim=1)
+        
+        if num_patches == self.patch_embedding.num_patches:
+            x += self.positional_embedding
+                
+        x = self.dropout_layer(x)
+        x = self.transformer_encoder(x)
+        x= x[:,0,:] 
+        prediction=self.prediction_mlp(x)
+        return prediction
+
+class PatchEmbedding(nn.Module):
+    def __init__(
+            self, 
+            img_size: Tuple[int],
+            in_channels: int, 
+            patch_size: int, 
+            embed_dim: int, 
+            freq: bool
+    ) -> None:
+        
+        super(PatchEmbedding, self).__init__()
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.in_channels = in_channels
+        self.img_size = img_size
+        self.freq = freq
+        self.num_patches = (img_size[0] // patch_size) * (img_size[1] // patch_size) + 1
+        self.projection = nn.Linear(patch_size * patch_size * in_channels, embed_dim)
+
+    def forward(self,x):
+        if self.freq == False:
+            batch_size = x.size(0)
+            x = x.unfold(2, self.patch_size, self.patch_size).unfold(3, self.patch_size, self.patch_size)
+            x = x.permute(0, 2, 3, 1, 4, 5).contiguous()
+            x = x.view(batch_size, -1, self.patch_size * self.patch_size * self.in_channels)
```

Comment:
You can use `einops.rearrange` to replace these four lines with:
 `rearrange('b c (h p1) (w p2) -> b (h w) (p1 p2 c)', p1 = self.patch_size[0], p2 = self.patch_size[1])`